### PR TITLE
RHCLOUD-31878 Updated error message when trying to add duplicate group

### PR DESCRIPTION
**Pull request by EvanCasey13**
## Link(s) to Jira
https://issues.redhat.com/browse/RHCLOUD-31878

## Description of Intent of Change(s)
Updated the Format of the message returned when trying to add a duplicate group to a tenant

## Local Testing
How can the feature be exercised?
How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [x] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/utils.py [hunk 1]
```diff
@@ -279,6 +279,12 @@ def validate_group_name(name):
         raise serializers.ValidationError({key: _(message)})
 
 
+def check_duplicate_entry(grp_name, tenant_id):
+    """Check for duplicate group per tenant."""
+    if Group.objects.filter(tenant=tenant_id).filter(name=grp_name):
+        raise serializers.ValidationError({"Group" + " " + grp_name + " " + "already exists": "Group already exists"})
+
+
```
Comment (petracihalova) [inline rbac/management/utils.py:287]: the `source` in the response should be in this case only `group` and message with the duplicate group name as `detail` property, so the response can look like this
```
 {
    "errors": [
        {
            "detail": "Group 'ABC' already exists.",
            "source": "Group ABC already exists",
            "status": "400"
        }
    ]
}
```

you can use [f strings](https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings) to increase the readability, something like `raise serializers.ValidationError({"group": f"Group '{grp_name}' already exists."})`

File: rbac/management/utils.py [hunk 2]
```diff
@@ -279,6 +279,12 @@ def validate_group_name(name):
         raise serializers.ValidationError({key: _(message)})
 
 
+def check_duplicate_entry(grp_name, tenant_id):
```
Comment (petracihalova) [inline rbac/management/utils.py:282]: good job 🎉  it works and that's always important, I have a few minor comments
I would be very happy for a more descriptive function name here
+
another option to consider could be to catch the error from db to eliminate the not necessary call when provided group name is not duplicate

Comment (EvanCasey13) [reply rbac/management/utils.py:282]: Cool I will update with your recommended changes

File: rbac/management/group/view.py [hunk 3]
```diff
@@ -256,6 +256,7 @@ def create(self, request, *args, **kwargs):
             }
         """
         validate_group_name(request.data.get("name"))
+        check_duplicate_entry(request.data.get("name"), self.request.tenant)
```
Comment (lpichler) [inline rbac/management/group/view.py:259]: @EvanCasey13 is it possible to consider this [comment](https://github.com/RedHatInsights/insights-rbac/pull/1217#discussion_r1786771247)  ?

it looks that there are some benefits with that.

Comment (EvanCasey13) [reply rbac/management/group/view.py:259]: Yes I think I can catch IntegrityException when create occurs will go that route and remove the function if it works

File: rbac/management/utils.py [hunk 4]
```diff
@@ -279,6 +279,12 @@ def validate_group_name(name):
...
+    """Check for duplicate group per tenant."""
+    if Group.objects.filter(tenant=tenant_id).filter(name=grp_name):
```
[the remark this answers is not in the document]
Comment (petracihalova) [reply rbac/management/utils.py]: closing, is no longer relevant

File: rbac/management/group/view.py [hunk 5]
```diff
@@ -256,7 +257,18 @@ def create(self, request, *args, **kwargs):
             }
         """
         validate_group_name(request.data.get("name"))
-        create_group = super().create(request=request, args=args, kwargs=kwargs)
+        try:
+            create_group = super().create(request=request, args=args, kwargs=kwargs)
+        except IntegrityError as e:
+            if "unique constraint" in str(e.args):
```
[the remark this answers is not in the document]
Comment (petracihalova) [reply rbac/management/group/view.py:263]: if I understand your intention properly, this is actually not working as you expected, the second exception will not be caught (please try to add a test for this case) + check the status code that is returned ... you can for example use something like `raise serializers.ValidationError({"group": f"Group '{grp_name}' already exists."})` that will solve the problem with the status code for you without additional settins

File: tests/management/group/test_view.py [hunk 6]
```diff
@@ -5376,3 +5376,33 @@ def test_manipulate_with_group_with_rbac_read_permission(self):
 
         response = client.put(url, request_body, format="json", **self.headers_service_account_principal)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_duplicate_entry_message(self):
```
Comment (petracihalova) [inline tests/management/group/test_view.py:5380]: please test the `response.status_code` too

**Comment by petracihalova** [comment]
hi Evan it looks nice 👍 I have one tiny comment ... now the response looks like this
```
{
    "errors": [
        {
            "detail": "A group with the name 'ABC3' exists for this tenant",
            "source": "detail",
            "status": "400"
        }
    ]
}
```

so we need to update the value for the `"source"` to be `"group"`

you can do it by changing the key value in the dict here
```
raise serializers.ValidationError({"group": f"A group with the name '{request.data.get('name')}' exists for this tenant"})
```

**Review by lpichler — APPROVED** [review]

**Pull request merged**